Alerting: Add rule_limits to rule list requests

alexander-akhmetov opened the pull request:
**What is this feature?**

Adds support for the `rule_limit` parameter in the Alerting frontend calls to the `/api/prometheus/grafana/api/v1/rules`. The parameter is conditionally used only when client-side filtering is not needed.


Part of https://github.com/grafana/alerting-squad/issues/1270

alexander-akhmetov committed:
Alerting: Add rule_limits to rule list requests

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit

konrad147 committed:
Unify pagination limits calculation for GMA and DMA rules

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit

konrad147 committed:
Fix limits, add tests

gillesdemey commented on public/app/features/alerting/unified/rule-list/hooks/datasourceFilter.ts:
```diff
@@ -22,6 +22,27 @@ import {
+/**
+ * Determines if client-side filtering is needed for data source-managed rules.
+ */
+export function hasDatasourceFilters(filterState: Partial<RulesFilter>): boolean {
```
nit: this function name seems confusing / unrelated to data sources being included in the filter :)

Perhaps `shouldPerformClientSideFiltering()` or similar ?

gillesdemey commented on public/app/features/alerting/unified/rule-list/paginationLimits.ts:
```diff
@@ -6,10 +13,36 @@ export const FILTERED_GROUPS_SMALL_API_PAGE_SIZE = 100;
+
+export function getFilteredRulesLimits(filterState: RulesFilter): FetchGroupsLimitOptions {
+  return {
+    gmaLimit: getGrafanaFilterLimits(filterState),
```
nit: prefer longer property names for clarity :)

gillesdemey reviewed (approved):
LGTM couple of nits / questions

konrad147 committed:
Alerting: Rename filter functions and limit properties for clarity

- hasClientSideFilters → hasGrafanaClientSideFilters
- hasDatasourceFilters → hasDatasourceClientSideFilters
- gmaLimit → grafanaManagedLimit
- dmaLimit → datasourceManagedLimit

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit

konrad147 commented:
/deploy-to-hg

konrad147 reviewed (approved):
LGTM!

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alexander-akhmetov/fe-rule-limit
